core/src/main/scala/org/apache/spark/ml/source/image/PatchedImageFileFormat.scala

```diff
@@ -62,15 +64,16 @@ class PatchedImageFileFormat extends ImageFileFormat with Serializable with Logg
   }
 
   //This is needed due to a multi-threading bug n the jvm
+  @annotation.tailrec
   private def catchFlakiness[T](times: Int)(f: => Option[T]): Option[T] = {
-    try {
+    Try {
       f
-    } catch {
-      case e: NullPointerException if times >= 1 =>
-        logWarning("caught null pointer exception due to jvm bug", e)
+    } match {
+      case Success(x) => x
+      case Failure(exception: NullPointerException) if times >= 1 =>
+        logWarning("caught null pointer exception due to jvm bug", exception)
         catchFlakiness(times - 1)(f)
-      case _: Exception =>
-        None
+      case _ => None
     }
   }
```

Comment:
Not related to the bug fix, but changed the function into functional style and add tailrec annotation.